internal/pkg/deploy/upload/asset/asset.go

```diff
@@ -71,25 +94,61 @@ func walkFn(source, dest string, recursive bool, reader afero.Fs, upload UploadF
 		if !ok {
 			return nil
 		}
-		file, err := reader.Open(path)
+
+		hash := sha256.New()
+		buf := &bytes.Buffer{}
+		file, err := u.FS.Open(path)
 		if err != nil {
-			return fmt.Errorf("open file on path %q: %w", path, err)
+			return fmt.Errorf("open %q: %w", path, err)
 		}
 		defer file.Close()
-		fileRel, err := filepath.Rel(source, path)
+
+		_, err = io.Copy(io.MultiWriter(buf, hash), file)
 		if err != nil {
-			return fmt.Errorf("get relative path for %q against %q: %w", path, source, err)
+			return fmt.Errorf("copy %q: %w", path, err)
 		}
-		key := filepath.Join(dest, fileRel)
-		url, err := upload(key, file)
-		if err != nil {
-			return fmt.Errorf("upload file %q to destination %q: %w", path, key, err)
+
+		asset := Cached{
+			LocalPath:       path,
+			Content:         buf,
+			CachePath:       filepath.Join(u.CachePathPrefix, hex.EncodeToString(hash.Sum(nil))),
+			CacheBucket:     u.CacheBucket,
+			DestinationPath: destPath,
+		}
+
+		if sourcePath == path && destPath == "" {
+			asset.DestinationPath = info.Name()
+		} else if sourcePath != path {
+			rel, err := filepath.Rel(sourcePath, path)
+			if err != nil {
+				return fmt.Errorf("get relative path for %q against %q: %w", path, sourcePath, err)
+			}
+
+			asset.DestinationPath = filepath.Join(destPath, rel)
```

Comment:
Is this because we can get rid of https://github.com/aws/copilot-cli/pull/4722/files#diff-befc0e16cf8c6a7ad74caee13773fd4b095ba8bea5170dd1e39f6e3384e0667fL35-L46 ?

Reply:
the relative path was the same previously, but yeah, with this we can get rid of the other section